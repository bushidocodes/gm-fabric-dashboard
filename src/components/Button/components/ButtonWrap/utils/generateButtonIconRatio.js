/**
 * Helper function that maps button sizes to numbers representing icon ratios
 *
 * @param {string} size
 * @returns number
 */
function mapSizeToButtonIconRatio(size) {
  switch (size) {
    case "xs":
      return 0.5;
    case "sm":
      return 0.75;
    case "lg":
      return 1.25;
    case "xl":
      return 1.5;
    case "normal":
    default:
      return 1;
  }
}

/**
 * Utility function that generates a CSS mixin containing button icon rulesets
 *
 * @param {string} size
 * @returns string
 */
function generateButtonIconRatio(size) {
  const viewboxSize = 24; // default viewbox size 24px
  const ratio = mapSizeToButtonIconRatio(size); // default ratio 1
  const viewboxDimensions = viewboxSize * ratio;
  return `> svg { width: ${/* sc-value */ viewboxDimensions}px; 
  height: ${/* sc-value */ viewboxDimensions}px; }`;
}

export default generateButtonIconRatio;
