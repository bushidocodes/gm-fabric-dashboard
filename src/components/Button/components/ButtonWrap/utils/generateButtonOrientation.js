import { spacingScale } from "style/styleFunctions";

/**
 * Utility function that generates a CSS mixin containing button orientation rulesets
 *
 * @param {string} [alignment="horizontal"]
 * @param {number} [spacingBase=0.5]
 * @returns string
 */
function generateButtonOrientation(
  alignment = "horizontal",
  spacingBase = 0.5
) {
  if (alignment === "horizontal") {
    return `
      flex-direction: row;
      > :not(:first-child) {
        margin-left: ${/* sc-value */ spacingScale(spacingBase)};
        flex-basis: auto;
      }
    `;
  } else if (alignment === "vertical") {
    return `
      flex-direction: column;
      padding-top: ${/* sc-value */ spacingScale(spacingBase)};
      > :not(:first-child) {
        margin-top: ${/* sc-value */ spacingBase / 2}px;
      }
    `;
  }
}

export default generateButtonOrientation;
