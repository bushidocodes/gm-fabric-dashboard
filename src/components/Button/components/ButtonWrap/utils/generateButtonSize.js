import { spacingScale } from "style/styleFunctions";

/**
 * Utility function that transforms an object containing different attributes
 * into a block of CSS attributes dealing with button size
 *
 * @param {any} {
 *   fontSizeMultiplier = 1,
 *   horizontalPaddingMultiplier = 1,
 *   verticalPaddingMultiplier = 1,
 *   textIsUppercase = false
 * }
 * @returns string
 */
function _buttonSizeStyleBlockGenerator({
  fontSizeMultiplier = 1,
  horizontalPaddingMultiplier = 1,
  verticalPaddingMultiplier = 1,
  textIsUppercase = false
}) {
  const fontSizeBase = 14;
  const borderRadiusBase = 3;
  const contentSpacingBase = parseInt(spacingScale(0.5), 10);
  return `
    font-size: ${Math.ceil(fontSizeBase * fontSizeMultiplier)}px;
    border-radius: ${borderRadiusBase}px;
    padding: ${contentSpacingBase *
      horizontalPaddingMultiplier}px ${contentSpacingBase *
    verticalPaddingMultiplier}px;
    ${textIsUppercase ? "text-transform: uppercase;" : ""}
  `;
}

/**
 * Utility function that renders a block of CSS attributes for a
 * particular string (xs, sm, normal, lg, xl)
 *
 * @param {string} size
 * @returns
 */
function generateButtonSize(size) {
  switch (size) {
    case "xs": {
      return _buttonSizeStyleBlockGenerator({
        fontSizeMultiplier: 0.5,
        horizontalPaddingMultiplier: 0.25,
        verticalPaddingMultiplier: 1,
        textIsUppercase: true
      });
    }
    case "sm": {
      return _buttonSizeStyleBlockGenerator({
        fontSizeMultiplier: 0.75,
        horizontalPaddingMultiplier: 0.25,
        verticalPaddingMultiplier: 1,
        textIsUppercase: true
      });
    }
    case "lg": {
      return _buttonSizeStyleBlockGenerator({
        fontSizeMultiplier: 1.25,
        horizontalPaddingMultiplier: 0.5,
        verticalPaddingMultiplier: 1,
        textIsUppercase: false
      });
    }
    case "xl": {
      return _buttonSizeStyleBlockGenerator({
        fontSizeMultiplier: 1.5,
        horizontalPaddingMultiplier: 0.5,
        verticalPaddingMultiplier: 1,
        textIsUppercase: false
      });
    }
    default:
    case "normal": {
      return _buttonSizeStyleBlockGenerator({
        fontSizeMultiplier: 1,
        horizontalPaddingMultiplier: 0.35,
        verticalPaddingMultiplier: 1,
        textIsUppercase: false
      });
    }
  }
}

export default generateButtonSize;
