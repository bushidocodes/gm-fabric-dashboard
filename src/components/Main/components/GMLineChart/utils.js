// importing a single function instead of the whole _
import isEmpty from "lodash/isEmpty";

/**
 * Converts timeseries with time stamps to a string of values without timestamps
 *
 * @param {array} array a timeseries array with time stamps
 * @returns {array}     an array of values
 */
export function numericalTimeSeriesFunc(array) {
  if (array instanceof Array === true && !isEmpty(array)) {
    return array[0].map(cell => cell[1]);
  } else {
    return 0;
  }
}

/**
 * Produces an average of array of values
 *
 * @param {array} array an array of values
 * @returns {number}
 */
export function average(array) {
  if (array instanceof Array === true && !isEmpty(array)) {
    return array.reduce((a, b) => a + b) / array.length;
  } else {
    return 0;
  }
}

/**
 * Produces a median of array of values
 * (if no middle value, average of 2 most central values is returned)
 *
 * @param {array} array an array of values
 * @returns {number}
 */
export function median(array) {
  if (array instanceof Array === true && !isEmpty(array)) {
    const len = array.length;
    let median = 0;
    array.sort();

    if (len % 2 === 0) {
      // if even, average of two middle numbers
      median = (array[len / 2 - 1] + array[len / 2]) / 2;
    } else {
      // if odd, middle number only
      median = array[(len - 1) / 2];
    }
    return median;
  } else {
    return 0;
  }
}

/**
 * Produces a mode(s) of array of values
 * if more than one mode, returns all applicable modes as well as their frequency
 *
 * @param {array} array an array of values
 * @returns {array}     an array that contains mode(s) and frequency value
 */
export function modes(array) {
  if (array instanceof Array === true && !isEmpty(array)) {
    let modes = [];
    let count = [];
    let i;
    let number;
    let maxIndex = 0;

    for (let i = 0; i < array.length; i++) {
      number = array[i];
      count[number] = (count[number] || 0) + 1;
      if (count[number] > maxIndex) {
        maxIndex = count[number];
      }
    }

    for (i in count)
      if (count.hasOwnProperty(i)) {
        if (count[i] === maxIndex) {
          modes.push(Number(i));
        }
      }

    modes.push({ frequency: maxIndex });

    if (maxIndex > 1) {
      return modes;
    } else {
      modes = "no value occurs more than once";
      return modes;
    }
  } else {
    return 0;
  }
}

/**
 * Produces a string that contains descriptive statistics of the graph
 *
 * @param {array} timeSeries an array of values
 * @param {string} title chart title
 * @returns {string}     returns a string with summary chart statistics
 */
export function screenReaderGraphDescription(timeSeries, title, intl) {
  const numericalTimeSeries = numericalTimeSeriesFunc(timeSeries);

  // We don't have negative values in our charts, thus average of 0 means that all values are 0, aka no meaningful data
  if (average(numericalTimeSeries) > 0) {
    let screenReaderGraphData = {
      median: median(numericalTimeSeries).toFixed(2),
      average: average(numericalTimeSeries).toFixed(2),
      mode: modes(numericalTimeSeries),
      max: numericalTimeSeries.sort()[numericalTimeSeries.length - 1],
      min: numericalTimeSeries.sort()[0],
      dataPoints: numericalTimeSeries.length,
      timeSeries: numericalTimeSeries,
      title
    };

    if (
      typeof screenReaderGraphData.mode !== "string" &&
      screenReaderGraphData.mode.length > 1
    ) {
      screenReaderGraphData = Object.assign({}, screenReaderGraphData, {
        mode: `Mode has frequency of ${
          screenReaderGraphData.mode.pop().frequency
        } and values of ${screenReaderGraphData.mode}`
      });
    }

    return intl.formatMessage(
      {
        id: "GMLineChart.screenReaderGraphDescription.withData",
        defaultMessage:
          "A tabular representation of the {title} chart data: median {median} average {average} mode {mode} maximum {max} minimum {min} number of observations {dataPoints} complete data time series follows {timeSeries}",
        description: "Screen reader description for GMLineChart"
      },
      screenReaderGraphData
    );
  } else {
    return intl.formatMessage({
      id: "GMLineChart.screenReaderGraphDescription.noData",
      defaultMessage: `The average for currently displayed data is equal to 0.`,
      description: "Screen reader description for GMLineChart"
    });
  }
}
