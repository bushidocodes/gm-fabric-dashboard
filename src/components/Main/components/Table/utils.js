import React from "react";
import TableColHeader from "components/Main/components/TableColHeader";
import TableColLatencyHeader from "components/Main/components/TableColLatencyHeader";

/** function getTableHeaders
 * takes Table type and returns appropriate headers with styling
 * for Function/Routes Table or GMServiceTable (Instances Table)
 * @param {object} items
 * @returns {React.element}
 */
export const getTableHeaders = type => {
  let headerNames = [];
  if (type === "Instance") headerNames = [type, "Uptime"];
  if (type === "Route" || type === "Function")
    headerNames = [type, "Requests/sec", "Requests", "Error %", "Latency"];
  let styleProps = {};

  // apply greater flex-basis for first column (20%)
  // requests/sec is a sparkline header and set to align left
  // set text-align to right for additional text columns
  return headerNames.map((header, headerIdx) => {
    styleProps = {};
    if (headerIdx === 0) styleProps = { flex: "1 1 20%" };
    if (headerIdx !== 0 && header !== "Requests/sec")
      styleProps = { textAlign: "right" };

    if (header === "Latency") {
      return <TableColLatencyHeader key={header} />;
    } else {
      return (
        <TableColHeader style={styleProps} key={header}>
          {header}
        </TableColHeader>
      );
    }
  });
};

// simple helper that evaluates and returns route or func props
export const getItem = (route, func) => {
  if (route !== undefined) return route;
  if (func !== undefined) return func;
};
