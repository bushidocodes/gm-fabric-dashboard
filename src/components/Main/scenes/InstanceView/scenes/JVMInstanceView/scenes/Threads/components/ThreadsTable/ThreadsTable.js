import { PropTypes } from "prop-types";
import React from "react";
import _ from "lodash";

import ThreadsTableHeader from "./components/ThreadsTableHeader";
import ThreadsList from "./components/ThreadsList";
import ThreadsTableStatusHeader from "./components/ThreadsTableStatusHeader";
import TableColHeader from "components/Main/components/TableColHeader";
import TableDisplay from "components/Main/components/TableDisplay";
import { threadStates } from "utils/constants";

ThreadsTable.propTypes = {
  filteredThreadData: PropTypes.array,
  groupByAttribute: PropTypes.string
};

ThreadsTable.defaultProps = {
  filteredThreadData: [],
  groupByAttribute: "none"
};

/**
 * Renders a table and appropriate headers based on the provided thread data
 * @export
 * @param {Object[]} { filteredThreadData = [] }
 * @param {string} { groupByAttribute = "none" }
 * @returns JSX.Element
 */

export default function ThreadsTable({ filteredThreadData, groupByAttribute }) {
  // If we are grouping threads by state, we need to render headers

  if (groupByAttribute === "state") {
    // Add a header property to each thread object
    // based on its state so that they can be grouped
    const mappedThreads = filteredThreadData.map(thread => {
      return {
        ...thread,
        header: getHeader(thread, groupByAttribute)
      };
    });

    // Create an object with keys matching headers
    // and values being an array of corresponding thread objects
    const dataGroupedByHeader = _.groupBy(
      mappedThreads,
      thread => thread.header
    );

    // Loop through thread states to preserve the order stopped, idle, active
    // only include headers that have corresponding threads
    let headers = threadStates.filter(
      header => !_.isEmpty(dataGroupedByHeader[header])
    );

    return (
      <TableDisplay>
        <ThreadsTableHeader />
        {headers.map((header, idx) => [
          <ThreadsTableStatusHeader key={`header|${header}|${idx}`}>
            <TableColHeader>{header}</TableColHeader>
          </ThreadsTableStatusHeader>,
          <ThreadsList
            threads={dataGroupedByHeader[header]}
            key={`list|${header}|${idx}`}
          />
        ])}
      </TableDisplay>
    );
  } else {
    return (
      <TableDisplay>
        <ThreadsTableHeader />
        <ThreadsList threads={filteredThreadData} />
      </TableDisplay>
    );
  }
}

function getHeader(thread, groupByAttribute) {
  switch (thread.state) {
    case "RUNNABLE":
      return "Active";
    case "WAITING":
    case "TIMED_WAITING":
      return "Idle";
    case "TERMINATED":
    case "BLOCKED":
    case "NEW":
      return "Stopped";
    case "None":
    default:
      return "none";
  }
}
