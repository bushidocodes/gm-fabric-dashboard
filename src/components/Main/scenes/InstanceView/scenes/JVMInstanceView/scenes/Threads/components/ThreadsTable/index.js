import ThreadsTable from "./ThreadsTable";
export default ThreadsTable;
