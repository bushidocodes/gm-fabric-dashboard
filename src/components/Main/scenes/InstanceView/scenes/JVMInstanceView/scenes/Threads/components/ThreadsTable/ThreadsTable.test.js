import React from "react";

import configureStore from "redux-mock-store";
import { renderWithIntl } from "utils/i18nTesting";
import mockState from "json/mockReduxState";
import ThreadsTable from "./index";

const mockStore = configureStore()(mockState);

describe("ThreadsTable component", () => {
  test("matches the snapshot", () => {
    expect(
      renderWithIntl(<ThreadsTable store={mockStore} />)
    ).toMatchSnapshot();
  });
});
