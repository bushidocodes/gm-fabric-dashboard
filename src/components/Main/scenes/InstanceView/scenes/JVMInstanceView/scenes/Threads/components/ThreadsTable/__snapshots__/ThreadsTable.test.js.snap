// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThreadsTable component Matched the snapshot 1`] = `
.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: rgb(255,255,255);
  box-shadow: inset 0 -1px hsl(0,0%,92%);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  font-weight: 400;
  overflow: visible;
  width: 100%;
}

.c3 {
  -webkit-flex: 1 1 12%;
  -ms-flex: 1 1 12%;
  flex: 1 1 12%;
  text-align: left;
  padding-right: 16px;
  padding-top: 7px;
  padding-bottom: 7px;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c3:first-child {
  padding-left: 16px;
}

.c2 {
  -webkit-flex: 1 1 12%;
  -ms-flex: 1 1 12%;
  flex: 1 1 12%;
  text-align: left;
  padding-right: 16px;
  padding-top: 7px;
  padding-bottom: 7px;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  -webkit-flex: 0 1 5em;
  -ms-flex: 0 1 5em;
  flex: 0 1 5em;
}

.c2:first-child {
  padding-left: 16px;
}

.c4 {
  display: block;
  list-style: none;
  margin: 0;
  padding: 0;
}

.c0 {
  display: block;
  font-family: Rubik,-apple-system,BlinkMacSystemFont,"Segoe UI","Roboto","Oxygen","Ubuntu","Cantarell","Fira Sans","Droid Sans","Helvetica Neue",sans-serif;
  margin-left: auto;
  margin-right: auto;
  max-width: 1300px;
  width: 100%;
}

<ThreadsTable
  filteredThreadData={Array []}
  groupByAttribute="none"
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <styled.div>
    <div
      className="c0"
    >
      <ThreadsTableHeader>
        <styled.div>
          <div
            className="c1"
          >
            <styled.div
              style={
                Object {
                  "textAlign": "center",
                }
              }
            >
              <div
                className="c2"
                style={
                  Object {
                    "textAlign": "center",
                  }
                }
              >
                State
              </div>
            </styled.div>
            <styled.div
              style={
                Object {
                  "textAlign": "center",
                }
              }
            >
              <div
                className="c2"
                style={
                  Object {
                    "textAlign": "center",
                  }
                }
              >
                ID
              </div>
            </styled.div>
            <styled.div>
              <div
                className="c3"
              >
                Name
              </div>
            </styled.div>
            <styled.div
              style={
                Object {
                  "flex": "0 1 7em",
                  "textAlign": "right",
                }
              }
            >
              <div
                className="c2"
                style={
                  Object {
                    "flex": "0 1 7em",
                    "textAlign": "right",
                  }
                }
              >
                Daemon
              </div>
            </styled.div>
            <styled.div
              style={
                Object {
                  "flex": "0 1 7em",
                  "textAlign": "right",
                }
              }
            >
              <div
                className="c2"
                style={
                  Object {
                    "flex": "0 1 7em",
                    "textAlign": "right",
                  }
                }
              >
                Priority
              </div>
            </styled.div>
            <styled.div>
              <div
                className="c2"
              />
            </styled.div>
          </div>
        </styled.div>
      </ThreadsTableHeader>
      <ThreadsList
        threads={Array []}
      >
        <styled.ol>
          <ol
            className="c4"
          />
        </styled.ol>
      </ThreadsList>
    </div>
  </styled.div>
</ThreadsTable>
`;
